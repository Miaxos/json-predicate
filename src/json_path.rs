use json_pointer::{JsonPointer, ParseError};
use serde::{Deserialize, Deserializer, Serialize};
use std::ops::Add;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSONPath(JsonPointer<String, Vec<String>>);

impl Default for JSONPath {
    fn default() -> Self {
        Self(JsonPointer::new(Vec::new()))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum JSONPathError {
    /// An invalid escape sequence was encountered, either a `~` escape or a
    /// `%` escape.
    #[error("An invalid escape sequence was encountered, either a `~` escape or a `%` escape.")]
    InvalidEscape,
    /// An error caused by not having a leading slash on the JSON pointer.
    ///
    /// For example, the string `a/b/c` is not a valid JSON pointer, while
    /// `/a/b/c` is.
    #[error("An error caused by not having a leading slash on the JSON pointer.")]
    NoLeadingSlash,
}

impl JSONPath {
    pub fn new<S: AsRef<str>>(value: S) -> Result<JSONPath, JSONPathError> {
        let ptr = value.as_ref().parse::<JsonPointer<_, _>>();
        match ptr {
            Ok(ptr) => Ok(JSONPath(ptr)),
            Err(ParseError::NoLeadingSlash) => Err(JSONPathError::NoLeadingSlash),
            Err(ParseError::InvalidEscape(_)) => Err(JSONPathError::InvalidEscape),
        }
    }

    pub fn empty() -> Self {
        Self::new("/").expect("Can't fail")
    }

    pub fn take(self) -> JsonPointer<String, Vec<String>> {
        self.0
    }
}

impl Serialize for JSONPath {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let val = self.0.to_string();
        Ok(serializer.serialize_str(&val)?)
    }
}

impl Add for JSONPath {
    type Output = Self;

    fn add(self, other: Self) -> Self::Output {
        // TODO: Fix it because it's an unwrap that shouldn't be here as we should be sure that it
        // works.
        Self::new(format!("{}{}", self.0.to_string(), other.0.to_string())).unwrap()
    }
}

impl<'de> Deserialize<'de> for JSONPath {
    fn deserialize<D>(deserializer: D) -> Result<JSONPath, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let ptr = s
            .parse::<JsonPointer<_, _>>()
            .map_err(|e| serde::de::Error::custom(format!("{:?}", e)))?;
        Ok(JSONPath(ptr))
    }
}
