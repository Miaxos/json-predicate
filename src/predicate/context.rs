use std::error::Error;

use crate::json_path::{JSONPath, JSONPathError};

/// The context to run the evaluation, you can have an empty context with
/// `PredicateContext::default()`.
#[derive(Default, Clone, Debug)]
pub struct PredicateContext {
    location: Option<JSONPath>,
}

#[derive(Debug, thiserror::Error)]
pub enum PredicateContextError {
    #[error("{0}")]
    Deserialize(#[from] Box<dyn Error>),
    #[error("{0}")]
    JSONPath(#[from] JSONPathError),
}

impl From<JSONPath> for PredicateContext {
    fn from(value: JSONPath) -> Self {
        Self {
            location: Some(value),
        }
    }
}

impl From<Option<JSONPath>> for PredicateContext {
    fn from(value: Option<JSONPath>) -> Self {
        Self { location: value }
    }
}

impl PredicateContext {
    pub fn new(path: String) -> Result<Self, PredicateContextError> {
        let location = JSONPath::new(path)?;

        Ok(Self {
            location: Some(location),
        })
    }

    pub fn final_path(&self, path: &Option<JSONPath>) -> Option<JSONPath> {
        // TODO: Remove cloning
        match (self.location.clone(), path.clone()) {
            (None, None) => None,
            (Some(path), None) | (None, Some(path)) => Some(path),
            (Some(root), Some(path)) => Some(root + path),
        }
    }
}
