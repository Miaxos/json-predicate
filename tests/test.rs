// Test adapted from https://github.com/MalcolmDwyer/json-predicate/blob/master/test/test.js

use json_predicate::builder::TestBuilder;
use json_predicate::context::PredicateContext;
use json_predicate::json_path::JSONPath;
use json_predicate::{FirstOrder, Predicate, PredicateImpl};

mod utils;
use utils::ENTRY;

#[test]
pub fn test_base_predicate_snapshot() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .ignore_case(false)
            .value("ABC")
            .build()
            .unwrap(),
    )
    .into();

    insta::assert_json_snapshot!(predicate);
}

#[test]
pub fn returns_true_for_matching_string_value() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .value("ABC")
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_false_for_stirng_mismatched_only_by_case_without_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .value("aBc")
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_true_for_string_mismatched_only_by_case_with_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .value("aBc")
            .ignore_case(true)
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_true_for_matching_shallow_object() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objX/objY").unwrap())
            .value(serde_json::json!({ "num2": 2 }))
            .ignore_case(true)
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_true_for_matching_deep_object() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objX").unwrap())
            .value(serde_json::json!({ "num1": 1, "stringAbc": "Abc", "objY": { "num2": 2 } }))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_true_for_matching_number_value() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objA/num2").unwrap())
            .value(2)
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_true_for_matching_array() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/arrayA").unwrap())
            .value(serde_json::json!(["a", "b", "c",]))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_false_for_superset_array() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/arrayA").unwrap())
            .value(serde_json::json!(["a", "b", "c", "d"]))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_false_for_subset_array() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/arrayA").unwrap())
            .value(serde_json::json!(["a", "b"]))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_false_for_string_disguised_as_array() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/arrayA").unwrap())
            .value("abcd")
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_false_for_array_disguised_as_a_string() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .value(serde_json::json!(["A", "B", "C"]))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_true_for_string_match_inside_object_honoring_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objX").unwrap())
            .ignore_case(true)
            .value(serde_json::json!({ "num1": 1, "stringAbc": "aBc", "objY": { "num2": 2 } }))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(result.unwrap());
}

#[test]
pub fn returns_false_for_string_inside_object_mismatched_only_by_case_without_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objX").unwrap())
            .value(serde_json::json!({ "num1": 1, "stringAbc": "aBc", "objY": { "num2": 2 } }))
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert!(!result.unwrap());
}

#[test]
pub fn returns_err_for_undefined_value() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objZZZ/objZZZZZZZZ").unwrap())
            .value("dzodeiojdoizej")
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_err());
    insta::assert_debug_snapshot!(result, @r###"
    Err(
        PredicateError {
            kind: JsonPointerIssue {
                issue: NoSuchKey(
                    "objZZZ",
                ),
            },
        },
    )
    "###);
}

#[test]
pub fn returns_err_for_undefined_value_with_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        TestBuilder::default()
            .path(JSONPath::new("/objZZZ/objZZZZZZZZ").unwrap())
            .ignore_case(true)
            .value("dzodeiojdoizej")
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_err());
    insta::assert_debug_snapshot!(result, @r###"
    Err(
        PredicateError {
            kind: JsonPointerIssue {
                issue: NoSuchKey(
                    "objZZZ",
                ),
            },
        },
    )
    "###);
}
