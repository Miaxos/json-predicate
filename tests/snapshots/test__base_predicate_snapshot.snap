---
source: tests/test.rs
expression: predicate

---
{
  "op": "test",
  "path": "/stringABC",
  "value": "ABC"
}
